### DISCUSSION TITLE
Best practices for api and lambda layers

### DISCUSSION THREAD

**Discussion by mteichtahl** [aws/aws-pdk#261] [Q&A]
I am trying to create a lambda layer that gets consumed by an API operation lambda. 

I have an apiInfra stack which is responsible for the infrastructure supporting the api (like the layer) and instantiating the api. 

The layer has its own dependencies. 

I have tried having the layer as a projen typescript project. However, the apiInfra stack needs  to take a dependency on the built layer. This fails as projen enforces a rootDir prohibiting access to files outside of the stack. 

The scenario here is to use typeorm across the entire application. This means the models etc are "common" and need to be referenced from multiple places such as the api lambda. 

Any suggestions on the right design pattern would be greatly appreciated.

**Comment by cogwirrel** [comment]
I think it makes sense to have the typeorm as its own typescript project if it's referenced by multiple other parts of your application. The `apiInfra` will need to take a dependency on the built project rather than importing source files directly from the other project though :). You'd need to make sure the typeorm files you need are all included in the build output, for example by copying them as a post compile step:

_.projenrc.ts_

```ts
const common = new TypeScriptProject({
  name: '@my-project/typeorm-common',
  ...
});
common.postCompileTask.exec('cp -r src/typeorm lib/');
```

For usage in your lambda - if you are using `NodejsFunction` you may not need to create a layer, and might instead be able to specify your typeorm package in the [nodeModules](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda_nodejs-readme.html#install-modules) to make the non-typescript files are available, eg:

```ts
new NodejsFunction(this, 'ApiLambda', {
  ...
  bundling: {
    nodeModules: ['@my-project/typeorm-common'],
  },
});
```

Hope that helps! :)

**Reply by mteichtahl** [reply]
Thanks. I'll give this a try and report back. Likely sometime next week.
